### PR TITLE
raftstore: use lru cache store local read delegate

### LINKED ISSUE #9469: raftstore: the invalid read delegates in local reader may not be removed forever
## Bug Report

Although https://github.com/tikv/tikv/pull/4765 introduce the `invalid` flag to remove the invalid `ReadDelegate`s on `LocalReader`, but the flag only be checked when handling a request to the `ReadDelegate`, if the `ReadDelegate` is marked as invalid and no incoming request to it, the `ReadDelegate` not be removed and consuming memory.

Although `mem::size_of::<ReadDelegate>()` is 122 bytes, but if this case happened the `ReadDelegate` will not be removed forever, and the memory occupied by these invalid `ReadDelegate` will only increase not decrease as TiKV running.

### PR DESCRIPTION
**Pull request by NingLin-P**
Signed-off-by: linning <linningde25@gmail.com>

### What problem does this PR solve?

Issue Number: close #9469 

### What is changed and how it works?

What's Changed:

Using `LruCache` instead of `HashMap` to store local `ReadDelegate`

### Release note 
- raftstore: use lru cache store local read delegate

### CODE REVIEW

**Comment by NingLin-P** [comment]
Do you have any suggestions about the capacity of the `LruCache`? @BusyJay @hicqu

File: components/raftstore/src/store/worker/read.rs [hunk 1]
```diff
@@ -373,7 +375,7 @@ where
             cache_read_id,
             store_id: Cell::new(None),
             metrics: Default::default(),
-            delegates: HashMap::default(),
+            delegates: LruCache::with_capacity(DELEGATE_LRU_CACHE_CAPACITY),
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/read.rs:378]: I suggest to use default. There can be no allocation at all for learners.

**Comment by BusyJay** [comment]
You can use `resize` to shrink the cache at every modification operations.

File: components/raftstore/src/store/worker/read.rs (hunk 1 above)
Comment (NingLin-P) [reply components/raftstore/src/store/worker/read.rs:378]: There is not `LruCache::default`,  underneath `LruCache::with_capacity` using `HashMap::default` to initialize, so it won't allocate large memory until it needs to.

**Comment by hicqu** [comment]
There are some `FIXME`s can be removed in `read.rs`.

**Comment by NingLin-P** [comment]
> You can use `resize` to shrink the cache at every modification operations.

If so, the problem still exists, local read will not call `delegate.remove` for those invalid delegates.

**Comment by NingLin-P** [comment]
> There are some `FIXME`s can be removed in `read.rs`.

The only `FIXME` is removed.

**Review by hicqu — APPROVED** [review]

**Comment by BusyJay** [comment]
> If so, the problem still exists, local read will not call delegate.remove for those invalid delegates.

What I mean is `cache.resize(meta.delegates.len())`.

**Comment by NingLin-P** [comment]
I also remove the `invalid` flag, PTAL again @hicqu @BusyJay

**Comment by hicqu** [comment]
/LGTM.

File: components/raftstore/src/store/worker/read.rs [hunk 2]
```diff
@@ -154,11 +153,8 @@ pub struct ReadDelegate {
 
 impl Drop for ReadDelegate {
     fn drop(&mut self) {
-        // `mark_invalid` if the source `ReadDelegate` is dropped
-        if self.track_ver.source {
-            self.mark_invalid();
-            self.track_ver.inc();
-        }
+        // call `inc` to notify the source `ReadDelegate` is dropped
+        self.track_ver.inc();
```
Comment (BusyJay) [inline components/raftstore/src/store/worker/read.rs:157]: Why not check `source` flag?

Comment (NingLin-P) [reply components/raftstore/src/store/worker/read.rs:157]: `inc` check the flag inside.

**Review by BusyJay — APPROVED** [review]

**Pull request merged**